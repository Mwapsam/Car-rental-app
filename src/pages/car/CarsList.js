import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { getCars } from '../../services/car.service';

const CarsList = () => {
  const dispatch = useDispatch();
  const { carsList, isCarsStored } = useSelector((state) => state.cars);
  useEffect(() => {
    if (!isCarsStored) { dispatch(getCars()); }
  }, []);

  return (
    <div>
      <h1>Cars</h1>
      {carsList && carsList.map((car) => (
        <Link to={`/cars/${car.id}`} key={car.id}>
          <div className="car-card">
            <img src={car.image} alt={car.name} />
            <h3>{car.name}</h3>
            <p>{car.description}</p>
          </div>
        </Link>
      ))}
    </div>
  );
};

export default CarsList;
