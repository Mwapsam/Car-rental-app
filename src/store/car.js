import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

// create thunk
export const getCars = createAsyncThunk('cars/getCars',
  async () => fetch('http://localhost:3000/api/v1/cars').then((res) => res.json()));

// create slice
export const carsSlice = createSlice({
  name: 'cars',
  initialState: {
    carsList: [],
    status: null,
  },
  extraReducers: {
    [getCars.pending]: (state) => {
      state.status = 'pending';
    },
    [getCars.fulfilled]: (state, action) => {
      state.carsList = action.payload;
      state.status = 'success';
    },
    [getCars.rejected]: (state) => {
      state.status = 'failed';
    },
  },
});

export default carsSlice.reducer;
