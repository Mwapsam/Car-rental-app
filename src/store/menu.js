/* eslint-disable consistent-return */
/* eslint-disable no-unused-vars */
import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'user',
  initialState: {
    active: false,
  },
  reducers: {

    menuState: (state) => {
      state.active = !state.active;
    },

  },
});

export default slice.reducer;

// Actions
const {
  menuState,
} = slice.actions;

export const updateMenuState = () => async (dispatch) => {
  dispatch(menuState());
};
