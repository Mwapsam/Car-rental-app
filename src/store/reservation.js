import { createSlice } from '@reduxjs/toolkit';
import getReservations, { CreateReservations } from '../services/reservation.service';

export const reservationsSlice = createSlice({
  name: 'reservations',
  initialState: {
    reservations: [],
    status: null,
  },
  extraReducers: {
    [getReservations.fulfilled]: (state, { payload }) => {
      state.reservations = payload;
      state.status = 'success';
    },
    [getReservations.pending]: (state) => {
      state.status = 'loading';
    },
    [getReservations.rejected]: (state) => {
      state.reservations = [];
      state.status = 'failed';
    },
    [CreateReservations.fulfilled]: (state, action) => {
      state.reservations.unshift(action.payload);
    },
  },
});
export default reservationsSlice.reducer;
