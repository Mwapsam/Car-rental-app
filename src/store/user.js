/* eslint-disable consistent-return */
/* eslint-disable no-unused-vars */
import { createSlice } from '@reduxjs/toolkit';
import UserService from '../services/user.service';
import LocalStorage from '../helpers/localStorage';

const initialUser = localStorage.getItem('user')
  ? JSON.parse(localStorage.getItem('user'))
  : null;

const slice = createSlice({
  name: 'user',
  initialState: {
    isLoggedIn: !!initialUser,
    data: initialUser,
  },
  reducers: {
    signupSuccess: (state, action) => {
      state.isLoggedIn = true;
      state.data = action.payload;
      localStorage.setItem('token', JSON.stringify(action.payload));
      window.history.pushState({}, '', '/');
      window.location.reload();
    },
    loginSuccess: (state, action) => {
      state.isLoggedIn = true;
      state.data = action.payload;
      localStorage.setItem('user', JSON.stringify(action.payload));
      window.history.pushState({}, '', '/');
      window.location.reload();
    },
    loginFailure: (state) => {
      state = initialUser;
    },
    signupFailure: (state) => {
      state = initialUser;
    },
    logoutUser: () => {
      LocalStorage.removeUser();
      window.history.pushState({}, '', '/login');
      window.location.reload();
    },
    setMessage: (state, action) => {
      state.error = action.payload;
    },
  },
});

export default slice.reducer;

// Actions
const {
  signupSuccess,
  loginSuccess,
  loginFailure,
  signupFailure,
  logoutUser,
  setMessage,
} = slice.actions;

export const login = (formData) => async (dispatch) => {
  try {
    const {
      data: { user, jwt, error },
    } = await UserService.login(formData);

    if (user && jwt) {
      const loggedInUser = { ...user, token: jwt };
      return dispatch(loginSuccess(loggedInUser));
    }
    if (error) {
      dispatch(setMessage(error));
      return dispatch(loginFailure());
    }
  } catch (e) {
    return dispatch(loginFailure());
  }
};
export const signup = (formData) => async (dispatch) => {
  try {
    const {
      data: { user, jwt, errors },
    } = await UserService.signup(formData);

    if (user && jwt) {
      const signedUpUser = { ...user, token: jwt };
      return dispatch(signupSuccess(signedUpUser));
    }
    if (errors) {
      return dispatch(signupFailure());
    }
  } catch (e) {
    // console
  }
};

export const logout = () => async (dispatch) => {
  dispatch(logoutUser());
};
