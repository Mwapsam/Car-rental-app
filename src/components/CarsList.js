import React, { useEffect } from "react";
import { useDispatch } from "react-redux";
import { getCars } from "../store/car";

const CarsList = () => {
    const dispatch = useDispatch()
    useEffect(() => {
        dispatch(getCars())
    }, [dispatch])

    return (
        <div>
            <h1>Cars</h1>
        </div>
    )
}

export default CarsList