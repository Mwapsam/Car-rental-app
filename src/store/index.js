import { configureStore } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';
import user from './user';

const reducer = combineReducers({
  user,
});
const store = configureStore({
  reducer,
});
export default store;
