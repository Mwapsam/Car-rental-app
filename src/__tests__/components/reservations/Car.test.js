import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../../store/index';
import CarsList from '../../../pages/car/CarsList';

it('should render Cars List component', () => {
const tree = renderer.create(
	<Provider store={store}>
		<CarsList />
	</Provider>
)
expect(tree).toMatchSnapshot();
});
