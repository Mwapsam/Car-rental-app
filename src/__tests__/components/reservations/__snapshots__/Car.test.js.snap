// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Cars List component 1`] = `
<div
  className="w-full"
>
  <h4
    className="font-black text-lg md:text-2xl uppercase p-0 m-0 "
  >
    Available Cars
  </h4>
  <div
    className="grid grid-cols-3 gap-4 grid-cols-1 md:grid-cols-2 lg:grid-cols-3 p-2"
  />
</div>
`;
